### ISSUE TITLE
Validation callbacks reset after AMP call

### ISSUE THREAD

**Issue by lorenzomammana** [ultralytics/ultralytics#1324]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Hello, I'm currently working with custom callbacks to integrate Mlflow logging, I managed to make it work properly for training, but my 'on_val_end' callback is never called once the model has finished training.

After debugging I've found out the reason and it's due to an improper call to _reset_callbacks() when the BaseTrainer.check_amp function is called.

My code is the following
```
        self.model = YOLO(model=self.config.model.model, task=self.config.model.task)

        if self.config.callbacks is not None:
            custom_callbacks = hydra.utils.instantiate(self.config.callbacks)
            if not isinstance(custom_callbacks, dict):
                raise ValueError("Custom callbacks must be a dictionary")

            for k, v in custom_callbacks.items():
                self.model.add_callback(k, v)

        self.model.overrides = {**self.config.model, **self.config.transforms}
        self.model.train()
```

custom_callbacks in the end is just a dictionary

```
 {
        "on_pretrain_routine_start": on_pretrain_routine_start,
        "on_fit_epoch_end": on_fit_epoch_end,
        "on_train_epoch_end": on_train_epoch_end,
        "on_val_end": on_val_end,
 }
```

On line 621 of yolo.engine.trainer.py there is this call
`assert amp_allclose(YOLO('yolov8n.pt'), im)`

This assert will create a new YOLO model, which triggers the _reset_callbacks() function, as such since the Validator is created later using default callbacks, the callbacks added from outside are removed!

### Environment

- Ultralytics YOLOv8.0.51 🚀 Python-3.9.15 torch-1.13.1+cu117 CUDA:0 (Tesla V100-DGXS-32GB, 32505MiB)
- OS: Ubuntu 20.04
- Python: 3.9

### Minimal Reproducible Example

```
model = YOLO("yolov8n.pt")
def evil_callback(validator):
    raise ValueError("This is skipped :)")

# This time evil_callback is skipped
model.add_callback("on_val_end", evil_callback)
model.train(**{"data": "path/to/data.yaml", "epochs": 1})

# Now the error is raised
model.add_callback("on_val_end", evil_callback)
model.val(**{"data": "path/to/data.yaml"})
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@lorenzomammana thanks for the bug report! AMP checks are a new feature intended to catch training NaNs before they start by disabling AMP if differences are found between normal inference and AMP inference. 

If you could help us by submitting a PR for a solution that works for you we can get started there. Thank you!

**Comment by lorenzomammana** [comment]
Hi @glenn-jocher I've opened a pr to fix this issue!

**Comment by glenn-jocher** [comment]
Removing TODO as resolved in #1345
